src/plugins/action_trim_responses/src/trimResponses.ts

```diff
@@ -0,0 +1,70 @@
+import databaseMethods from './databaseMethods'
+import isEqual from '@openmsupply/deep-comparison'
+
+interface Response {
+  id: number
+  code: string
+  value: { [key: string]: any }
+  comment?: string
+  decision?: string
+  time_created: any
+}
+
+interface ResponsesByCode {
+  [key: string]: Response[]
+}
+
+module.exports['trimResponses'] = async function (input: any, DBConnect: any) {
+  const db = databaseMethods(DBConnect)
+
+  console.log('Trimming unchanged duplicated responses...')
+
+  try {
+    const { applicationId, reviewId } = input
+    // Get ALL responses associated with application OR review
+    const responses = reviewId
+      ? await db.getAllReviewResponses(reviewId)
+      : await db.getAllApplicationResponses(applicationId)
+
+    // Create object of responses indexed by code, with response
+    // objects added to array for each code
+    const responsesByCode: ResponsesByCode = {}
+    responses.forEach((response: Response) => {
+      const { code } = response
+      if (reviewId) response.value = { comment: response.comment, decision: response.decision }
+      if (!(code in responsesByCode)) responsesByCode[code] = [response]
+      else responsesByCode[code].push(response)
+    })
+
+    const responsesToDelete: number[] = []
+
+    Object.values(responsesByCode).forEach((responseArray) => {
+      if (responseArray.length === 1) return
```

Comment:
do we currently trim 'undefined' responses ?

Reply:
Any response that has been duplicated, will be compared with its previous -- including 'undefined' responses. Is that what you meant?

Reply:
Say the reviewer didn't enter a decision, do we keep undefined decision ? Likewise for applications, do we trim 'undefined' responses (for optionals) ?

Reply:
I'm not sure, but I'm not getting how that's relevant for this particular action. What's being trimmed are duplicate responses, so the question of whether 'undefined' responses should be kept would need to be addressed at the original submission, right? 

I wonder if we have a slightly different view of what this action is required to do. Here's my understanding of where it fits into the process:

- A user (reviewer or applicant) gets "changes requested", and that becomes the status of their application/review.
- When they "START" their changes, duplicates are made for all their existing "current" (i.e. latest) responses, regardless of their state or value (undefined, etc.)
- They make some changes, and re-submit
- On re-submission, this is where **this Action** runs: from ALL responses associated with each question Code, the latest one of each is kept only IF it has changed from the immediately previous one, otherwise it's deleted and the previous one remains the "current"/latest.

Is that correct?

Reply:
Yep, I was thinking of a scenario where applicant or reviewer doesn't have a review response for particular element/application response. This should also be trimmed right ? (or does it happen in another action?)